File: javascript/node-postgres/src/example_preferred.js
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import assert from "node:assert";
import { AuroraDSQLPool } from "@aws/aurora-dsql-node-postgres-connector";

const NUM_CONCURRENT_QUERIES = 8;

function createPool(clusterEndpoint, user) {
  return new AuroraDSQLPool({
    host: clusterEndpoint,
    user: user,
    max: 10,
    idleTimeoutMillis: 30000,
    connectionTimeoutMillis: 10000,
    retry: { maxRetries: 5 },
  });
}

async function worker(pool, workerId) {
  const result = await pool.query("SELECT $1::int as worker_id", [workerId]);
  console.log(`Worker ${workerId} result: ${result.rows[0].worker_id}`);
  assert.strictEqual(result.rows[0].worker_id, workerId);
}

async function example() {
  const clusterEndpoint = process.env.CLUSTER_ENDPOINT;
  assert(clusterEndpoint, "CLUSTER_ENDPOINT environment variable is not set");
  const user = process.env.CLUSTER_USER;
  assert(user, "CLUSTER_USER environment variable is not set");

  const pool = createPool(clusterEndpoint, user);

  try {
    // Run concurrent queries using the connection pool
    const workers = [];
    for (let i = 1; i <= NUM_CONCURRENT_QUERIES; i++) {
      workers.push(worker(pool, i));
    }

    // Wait for all workers to complete
    await Promise.all(workers);

    console.log("Connection pool with concurrent connections exercised successfully");

    // Create table
    await pool.query(`CREATE TABLE IF NOT EXISTS owner (
      id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
      name VARCHAR(30) NOT NULL,
      city VARCHAR(80) NOT NULL,
      telephone VARCHAR(20)
    )`);

    // Transactional write with OCC retry
    await pool.transaction(async (client) => {
      await client.query(
        "INSERT INTO owner(name, city, telephone) VALUES($1, $2, $3)",
        ["John Doe", "Anytown", "555-555-1900"],
      );
    });

    // Verify the write
    const result = await pool.query("SELECT name, city FROM owner WHERE name = $1", ["John Doe"]);
    assert.strictEqual(result.rows[0].name, "John Doe");
    assert.strictEqual(result.rows[0].city, "Anytown");
    console.log(`Inserted: name=${result.rows[0].name}, city=${result.rows[0].city}`);

    // Clean up
    await pool.query("DELETE FROM owner WHERE name = $1", ["John Doe"]);

    console.log("Transactional write with OCC retry exercised successfully");
  } catch (error) {
    console.error(error);
    throw error;
  } finally {
    await pool.end();
  }
}

export { example };

File: rust/sqlx/src/example_preferred.rs
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use aurora_dsql_sqlx_connector::{txn, DsqlConnectOptions, OCCRetryExt};
use sqlx::postgres::PgPoolOptions;
use sqlx::{Executor, Row};

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let cluster_endpoint = std::env::var("CLUSTER_ENDPOINT")
        .expect("CLUSTER_ENDPOINT environment variable is not set");
    let cluster_user = std::env::var("CLUSTER_USER").unwrap_or_else(|_| "admin".to_string());

    let conn_str = format!("postgres://{}@{}/postgres", cluster_user, cluster_endpoint);

    // Admin users operate in the default "public" schema.
    // Non-admin users operate in a custom "myschema" schema.
    let schema = if cluster_user == "admin" {
        "public"
    } else {
        "myschema"
    };

    // Build config and create a pool with custom options.
    // connect_with() verifies connectivity and spawns a background token refresh task.
    let config = DsqlConnectOptions::from_connection_string(&conn_str)?;
    let schema_owned = schema.to_string();
    let mut pool = aurora_dsql_sqlx_connector::pool::connect_with(
        &config,
        PgPoolOptions::new()
            .max_connections(10)
            .after_connect(move |conn, _meta| {
                let schema = schema_owned.clone();
                Box::pin(async move {
                    let sql = format!("SET search_path = '{}'", schema);
                    conn.execute(sqlx::AssertSqlSafe(sql)).await?;
                    Ok(())
                })
            }),
    )
    .await?;

    // -- Concurrent read queries --
    let mut handles = Vec::new();
    for i in 0..5 {
        let pool = pool.clone();
        handles.push(tokio::spawn(async move {
            let row = sqlx::query("SELECT $1::int as value")
                .bind(i)
                .fetch_one(&pool)
                .await?;
            Ok::<i32, anyhow::Error>(row.get("value"))
        }));
    }

    for handle in handles {
        let result = handle.await??;
        println!("Worker result: {}", result);
    }

    println!("Concurrent pool operations completed successfully");

    // -- Setup table --
    pool.execute(
        "CREATE TABLE IF NOT EXISTS owner(
            id uuid NOT NULL DEFAULT gen_random_uuid(),
            name varchar(30) NOT NULL,
            city varchar(80) NOT NULL,
            PRIMARY KEY (id))",
    )
    .await?;

    // -- Transactional write WITH OCC retry (using trait and txn! macro) --
    pool.transaction_with_retry(None, |tx| {
        txn!({
            sqlx::query("INSERT INTO owner(name, city) VALUES($1, $2)")
                .bind("John Doe")
                .bind("Anytown")
                .execute(&mut **tx)
                .await?;
            Ok(())
        })
    })
    .await?;

    // Verify the write
    let row = sqlx::query("SELECT name, city FROM owner WHERE name = $1")
        .bind("John Doe")
        .fetch_one(&pool)
        .await?;

    let name: &str = row.get("name");
    let city: &str = row.get("city");
    println!("Inserted: name={}, city={}", name, city);

    // -- Transactional write WITHOUT OCC retry (opt-out) --
    // For operations that don't need retry, use sqlx directly
    let mut tx = pool.begin().await?;
    sqlx::query("DELETE FROM owner WHERE name = $1")
        .bind("John Doe")
        .execute(&mut *tx)
        .await?;
    tx.commit().await?;

    println!("Transactional write completed successfully");

    // Closing the pool stops the background refresh task.
    pool.close().await;

    Ok(())
}
